[KED-2891] Implement `spark.DeltaTable` dataset (#964)
quantumblacklabs/kedro: docs/source/11_tools_integration/01_pyspark.md

assert isinstance(df, pyspark.sql.DataFrame)
```

## Spark and Delta Lake interaction

[Delta Lake](https://delta.io/) is an open-source project that enables building a Lakehouse architecture on top of data lakes. It provides ACID transactions and unifies streaming and batch data processing on top of existing data lakes, such as S3, ADLS, GCS, and HDFS.
To setup PySpark with Delta Lake, have a look at [the recommendations in Delta Lake's documentation](https://docs.delta.io/latest/quick-start.html#python).

We recommend the following workflow, which makes use of the [Transcoding](../05_data/01_data_catalog.md) feature in Kedro.

```yaml
temperature:
  type: spark.SparkDataSet
  filepath: data/01_raw/data.csv
  file_format: csv
  load_args:
    header: True
    inferSchema: True
  save_args:
    sep: '|'
    header: True

weather@spark:
  type: spark.SparkDataSet

>   filepath: data/01_raw/data

Reviewer: ```suggestion
  filepath: s3:/my_bucket/03_primary/temperature
```

Reviewer 2: Is it not 
```suggestion
  filepath: s3a://my_bucket/03_primary/temperature
```
for spark?

  file_format: delta
  save_args:
    mode: "overwrite"
    df_writer:
      versionAsOf: 0

weather@delta:
  type: spark.DeltaTableDataSet

>   filepath: data/01_raw/data

Reviewer: ```suggestion
  filepath: s3:/my_bucket/03_primary/weather
```

Reviewer 2: ```suggestion
  filepath: s3a://my_bucket/03_primary/weather
```

```

The `DeltaTableDataSet` does not support `save()` operation, as the updates happen in place inside the node function, i.e. through `DeltaTable.update()`, `DeltaTable.delete()`, `DeltaTable.merge()`.

> > Note: `before_dataset_saved` hook will not run at the expected time, as the actual saving happens in the node through the DeltaTable operations.

Reviewer: ```suggestion

> Since the save operation happens within the `node` via the  DeltaTable API, the Kedro `before_dataset_saved` hook will not be triggered.
```

```python
Pipeline(
    [

>         node(func=..., inputs="temperature", outputs="weather@spark"),

Reviewer: so I think these need to be named and we need dummy implementation to show how "first_operation_complete" comes to be

Reviewer: Perhaps this image is good enough?
![image](https://user-images.githubusercontent.com/35801847/143282216-59ccd3f6-03c8-428a-9501-cea791060264.png)

Reviewer 2: Thanks, added!

>         node(func=..., inputs="temperature", outputs="weather@spark"),

Reviewer: ```suggestion
        node(func=process_barometer_data, inputs="temperature", outputs="weather@spark"),
```

>         node(func=..., inputs="weather@delta", outputs="first_operation_complete"),

Reviewer: ```suggestion
        node(func=update_meterological_state, inputs="weather@delta", outputs="first_operation_complete"),
```

        node(

>             func=...,

Reviewer: ```suggestion
            func=estimate_weather_trend,
```

            inputs=["first_operation_complete", "weather@delta"],
            outputs="second_operation_complete",
        ),
    ]
)
```

> `first_operation_complete` is a `MemoryDataSet` and it signals that a set of Delta operations is complete. This can be used as input to a downstream node, to preserve the shape of the DAG. Otherwise, if no downstream nodes need to run after this, the node can simply not return anything:

Reviewer: ```suggestion
`first_operation_complete` is a `MemoryDataSet` and it signals that any Delta operations which occur 'outside' the Kedro DAG are complete. This can be used as input to a downstream node, to preserve the shape of the DAG. Otherwise, if no downstream nodes need to run after this, the node can simply not return anything:
```

assert isinstance(df, pyspark.sql.DataFrame)
```

## Spark and Delta Lake interaction

[Delta Lake](https://delta.io/) is an open-source project that enables building a Lakehouse architecture on top of data lakes. It provides ACID transactions and unifies streaming and batch data processing on top of existing data lakes, such as S3, ADLS, GCS, and HDFS.
To setup PySpark with Delta Lake, have a look at [the recommendations in Delta Lake's documentation](https://docs.delta.io/latest/quick-start.html#python).

> We recommend the following workflow, which makes use of the [Transcoding](../05_data/01_data_catalog.md) feature in Kedro.

Reviewer 2: Added a few more sentences in the beginning. ✅

```yaml
temperature:
  type: spark.SparkDataSet

>   filepath: data/01_raw/data.csv

Reviewer: ```suggestion
  filepath: s3://my_bucket/01_raw/data.csv
```

  file_format: csv
  load_args:
    header: True
    inferSchema: True
  save_args:
    sep: '|'
    header: True

weather@spark:
  type: spark.SparkDataSet

>   filepath: s3a://my_bucket/03_primary/temperature

Reviewer: ```suggestion
  filepath: s3://my_bucket/03_primary/temperature
```

  file_format: delta
  save_args:
    mode: "overwrite"
    df_writer:
      versionAsOf: 0

weather@delta:
  type: spark.DeltaTableDataSet

>   filepath: s3a://my_bucket/03_primary/weather

Reviewer: ```suggestion
  filepath: s3://my_bucket/03_primary/weather
```

```

The `DeltaTableDataSet` does not support `save()` operation, as the updates happen in place inside the node function, i.e. through `DeltaTable.update()`, `DeltaTable.delete()`, `DeltaTable.merge()`.

> > Since the save operation happens within the `node` via the DeltaTable API, the Kedro `before_dataset_saved` hook will not be triggered.

Reviewer 2: Good point, it's the second option. I've rephrased the sentence, hopefully it clearer now.
